receiver/azuremonitorreceiver/scraper_batch.go

```diff
@@ -0,0 +1,560 @@
+	if err = s.loadCredentials(); err != nil {
+		return err
+	}
+
+	s.subscriptions = map[string]*azureSubscription{}
+	s.resources = map[string]map[string]*azureResource{}
+	s.regions = map[string]map[string]struct{}{}
+
+	// Initialize subscription ids from the config. Will be overridden if discovery is enabled anyway.
+	for _, id := range s.cfg.SubscriptionIDs {
+		s.loadSubscription(id)
+	}
+
+	return
+}
+
+func (s *azureBatchScraper) loadSubscription(id string) {
+	s.resources[id] = make(map[string]*azureResource)
+	s.subscriptions[id] = &azureSubscription{
+		SubscriptionID: id,
+	}
+	s.regions[id] = make(map[string]struct{})
+}
+
+func (s *azureBatchScraper) unloadSubscription(id string) {
+	delete(s.resources, id)
+	delete(s.subscriptions, id)
+	delete(s.regions, id)
+}
+
+// TODO: duplicate
+func (s *azureBatchScraper) loadCredentials() (err error) {
+	switch s.cfg.Authentication {
+	case defaultCredentials:
+		if s.cred, err = s.azDefaultCredentialsFunc(nil); err != nil {
+			return err
+		}
+	case servicePrincipal:
+		if s.cred, err = s.azIDCredentialsFunc(s.cfg.TenantID, s.cfg.ClientID, s.cfg.ClientSecret, nil); err != nil {
+			return err
+		}
+	case workloadIdentity:
+		if s.cred, err = s.azIDWorkloadFunc(nil); err != nil {
+			return err
+		}
+	case managedIdentity:
+		var options *azidentity.ManagedIdentityCredentialOptions
+		if s.cfg.ClientID != "" {
+			options = &azidentity.ManagedIdentityCredentialOptions{
+				ID: azidentity.ClientID(s.cfg.ClientID),
+			}
+		}
+		if s.cred, err = s.azManagedIdentityFunc(options); err != nil {
+			return err
+		}
+	default:
+		return fmt.Errorf("unknown authentication %v", s.cfg.Authentication)
+	}
+	return nil
+}
+
+func (s *azureBatchScraper) scrape(ctx context.Context) (pmetric.Metrics, error) {
+	s.getSubscriptions(ctx)
+
+	for subscriptionID, subcription := range s.subscriptions {
+		s.getResourcesAndTypes(ctx, subscriptionID)
+
+		resourceTypesWithDefinitions := make(chan string)
+		go func(subscriptionID string) {
+			defer close(resourceTypesWithDefinitions)
+			for resourceType := range s.resourceTypes[subscriptionID] {
+				s.getResourceMetricsDefinitionsByType(ctx, subscriptionID, resourceType)
+				resourceTypesWithDefinitions <- resourceType
+			}
+		}(subscriptionID)
+
+		var wg sync.WaitGroup
+		for resourceType := range resourceTypesWithDefinitions {
+			wg.Add(1)
+			go func(subscriptionID, resourceType string) {
+				defer wg.Done()
+				s.getBatchMetricsValues(ctx, subscriptionID, resourceType)
+			}(subscriptionID, resourceType)
+		}
+
+		wg.Wait()
+
+		// Once all metrics has been collected for one subscription, we move to the next.
+		// We need to keep it synchronous to have the subscription id in resource attributes and not metrics attributes.
+		// It can be revamped later if we need to parallelize more, but currently, resource emit is not thread safe.
+		rb := s.mb.NewResourceBuilder()
```

Comment:
Running each subscriptions synchronously might become a problem when tenant contains a lot of subscriptions. 
In our fork, we run each sub in parallel and have subscription id at data point level, but I prefer to keep it at resource attribute level. 

We need first to run it on our tenants and see if it takes longer than collection interval.

Reply:
Tested with 94 subscriptions discovered and it took largely less than 1m. On my laptop. Quite satisfying and reassuring.

Reply:
Wel ok it was for only one resource type 
```yaml
    services: [ "Microsoft.Storage/storageAccounts" ]
    use_batch_api: true
    discover_subscriptions: true
```

Reply:
I think it's acceptable to release it like that since it's not a migration but just a proposal of using a different API. 
I struggle to try big metrics amount locally but for sure we'll test in our environments and probably come back with fixes 😄